src/include/cephfs/types.h

```diff
@@ -520,6 +521,7 @@ struct inode_t {
 
   double export_ephemeral_random_pin = 0;
   bool export_ephemeral_distributed_pin = false;
+  std::string bal_rank_mask;
```

Comment:
So, probably the main concern that I have with this change is the increase in the size of the inode which would happen irrespective of the xattr being set.

What are your thoughts on persisting this in the xattr map @yongseokoh ?

Reply:
@vshankar There are downsides that may affect inode size. On the other hand, CephFS users have the advantage of being able to flexibly map different subtrees to multiple mds ranks depending on their use case. I think the benefits are great. Also, CephFS users are familiar with using xattr, so I think this method is convenient.

Reply:
That's fine - the xattr interface can remain as it is - just that we could benefit from not increasing the inode size if we can store the mask in xattr_map rather than in the inode.

Reply:
Do you mean xattr_map used in https://github.com/ceph/ceph/pull/36603? I will try to see if it can be employed.

Reply:
@vshankar 

I have a rough understanding of xattrs through ceph.mirror.dirty_snap_id xattr.

   ``` 
"xattrs": [
         {
             "key": "ceph.mirror.dirty_snap_id",
             "val": "7"
         }
     ]
```

Using xattrs seems like a good idea. However, it would be better for this proposed PR to be managed as the ceph.dir.* family. It seems to be easier to manage in the long term if it is managed in a similar way to the existing ceph.dir.pin and ceph.dir.pin.random. Then, please give your opinion.

Reply:
> Using xattrs seems like a good idea. However, it would be better for this proposed PR to be managed as the ceph.dir.* family. It seems to be easier to manage in the long term if it is managed in a similar way to the existing ceph.dir.pin and ceph.dir.pin.random. Then, please give your opinion.

You do have a point here. I was recommended to use xattrs mirror related xattrs for the sole reason of increasing the inode size. The added member in this change is

> std::string bal_rank_mask;

which when empty and encoded would add (I think) at least 4 bytes to the on-disk size (since the string length needs to be encoded before the actual string). @batrick What do you suggest?

Reply:
It would be trivial to add `bal_rank_mask` to the xattr map and add support in the client. As Venky mentioned, see this code:

https://github.com/ceph/ceph/blob/95455490bcdec81906f258e57dbb0fd3e1570e16/src/client/Client.cc#L13808-L13810

https://github.com/ceph/ceph/blob/95455490bcdec81906f258e57dbb0fd3e1570e16/src/client/Client.cc#L14324-L14337

It should be a trivial change. Please make this change @yongseokoh .

Reply:
@vshankar @batrick I will change it to use the xattrs map based on your feedback. (It was late because I was on vacation until yesterday. )

Reply:
@yongseokoh Sure thing - thx!

Reply:
@vshankar @batrick  I changed this PR to use xattrs map. I confirmed that ceph.dir.bal.mask xattr can be changed through setfattr/getfattr even without changing Client.cc. 
 https://github.com/ceph/ceph/pull/52373/commits/afa6c736b95a1286f9b6b165c5a69068cf8205bf#diff-277dc6e796ccecb6aa14c9357f7a86898d6ddcf4113c110a4e00e0a10f6fefa7R6427

 Do I have to change client.cc as mentioned by Patrick ?

 https://github.com/ceph/ceph/pull/52373#discussion_r1473082207

Reply:
> Do I have to change client.cc as mentioned by Patrick ?

I don't think that's required since the mirror_info xattr is a bit differently handled as it's a compound xattr.

Make sure relevant tests are added for validating get/set of the xattr.

Reply:
Yep, tests are included. https://github.com/ceph/ceph/pull/52373/files#diff-b6dd60ae517c65a7733ae0a977eab56298d8f227cbdb4cd0589b8685cbd9669a